### PR TITLE
FEATURE add config var for namespace mapping page->control

### PR DESCRIPTION
**Pull request by andrewandante**
Creates a config variable where you can define a mapping between your Pages namespace and your Controllers namespace.

A common desired pattern is to keep things properly organised by having `App\Pages\MyPageType` and `App\Control\MyPageTypeController` - but this isn't currently caught by the ✨ "magic" ✨ . This allows for a global setting, rather than the one-off afforded by `$controller_name`.

Opens the door for adding a default map in 5 - maybe auto-substituting Page => Control, but that's another debate to have.

### CODE REVIEW

File: code/Model/SiteTree.php [hunk 1]
```diff
@@ -2959,17 +2977,28 @@ public function getControllerName()
...
                 Deprecation::notice(
                     '5.0',
                     'Underscored controller class names are deprecated. Use "MyController" instead of "My_Controller".',
                     Deprecation::SCOPE_GLOBAL
                 );
-                $controller = $candidate;
-                break;
+                return $candidate;
+            } elseif ($namespaceMap !== null) {
+                foreach ($namespaceMap as $pageNamespace => $controllerNamespace) {
```
Comment (michalkleiner) [inline code/Model/SiteTree.php]: Add a check `$namespaceMap` is an associative array?

Comment (andrewandante) [reply code/Model/SiteTree.php]: Could do - it's likely to just fail on the `strpos` check though, which I think is acceptable. The config variable is called `namespace_mapping` after all, it does indicate an asssociative array should be set

Comment (michalkleiner) [reply code/Model/SiteTree.php]: It wouldn't get to the `strpos` check if it wasn't an array, failing with `Invalid argument supplied for foreach()`

Comment (andrewandante) [reply code/Model/SiteTree.php]: Ah well if it's not an `array` at all, yeah - I thought you meant it if it was a sequential array.

So basically you are suggesting:

```diff
- } elseif ($namespaceMap !== null) {
+ } elseif (is_array($namespaceMap) {
```

is that right?

Comment (michalkleiner) [reply code/Model/SiteTree.php]: Initially, I meant to check for an actual associative array as it's a map, but checking for an array as a minimum is a good-enough middle ground.

File: code/Model/SiteTree.php [hunk 2]
```diff
@@ -2959,17 +2977,28 @@ public function getControllerName()
...
                     '5.0',
                     'Underscored controller class names are deprecated. Use "MyController" instead of "My_Controller".',
                     Deprecation::SCOPE_GLOBAL
                 );
-                $controller = $candidate;
-                break;
+                return $candidate;
+            } elseif (is_array($namespaceMap)) {
+                foreach ($namespaceMap as $pageNamespace => $controllerNamespace) {
+                    if (strpos($class, $pageNamespace) === false) {
```
Comment (GuySartorelli) [inline code/Model/SiteTree.php:2991]:
```suggestion
                    if (strpos($class, $pageNamespace) !== 0) {
```
Bit of an edge case but we should probably make sure the mapping namespace isn't actually somewhere in the middle of the class name.

File: tests/php/Model/SiteTreeTest.php [hunk 3]
```diff
@@ -1622,6 +1624,13 @@ public function testGetControllerNameFromConfig()
         Config::inst()->update(Page::class, 'controller_name', 'This\\Is\\A\\New\\Controller');
         $class = new Page;
         $this->assertSame('This\\Is\\A\\New\\Controller', $class->getControllerName());
+
+        Config::inst()->update(SiteTree::class, 'namespace_mapping', [
+            'SilverStripe\\CMS\\Tests\\Page' => 'SilverStripe\\CMS\\Tests\\Controllers',
+        ]);
+
+        $namespacedSiteTree = new SiteTreeTest_NamespaceMapTest();
+        $this->assertSame(SiteTreeTest_NamespaceMapTestController::class, $namespacedSiteTree->getControllerName());
```
Comment (GuySartorelli) [inline tests/php/Model/SiteTreeTest.php:1633]: This test should really be in a new method.

**Review by GuySartorelli — CHANGES_REQUESTED** [review]
Couple of small changes to make, and some conflicts to resolve.
@andrewandante I know this has been sitting here for a long time without any action but if you're willing to make these changes I'd be happy to help get this across the line.

**Comment by andrewandante** [comment]
@GuySartorelli changes made and rebased - I seem to have a couple of fixes in there for some other tests, I don't remember what that is about but hopefully they are useful changes?? I can remove them if needed. I honestly don't remember why they are there 😅

**Comment by GuySartorelli** [comment]
@andrewandante I had intended to give those a proper look and see what is going on with them but I don't think I'll have time to.😅If you could remove those unrelated test changes that would be great, then I can merge on the merits of the new functionality and its related tests.

**Comment by andrewandante** [comment]
@GuySartorelli consider them removed - if the tests break I'll add them back 😉

**Review by GuySartorelli — APPROVED** [review]
Nice work!
Keeps the default behaviour, but gives some very flexible new configuration which allow developers to use naming conventions that may be more appropriate than the defaults.
